mammoth/distributed/communication.py

```diff
@@ -35,81 +36,48 @@ def broadcast_tensors(tensors, src=0, group=None):
             torch.distributed.broadcast(t, src, group=group)
 
 
-def only_ready_reduce_and_rescale_grads(named_parameters, group=None):
+def managed_reduce_and_rescale_grads(
+    named_parameters,
+    has_local_gradient: bool,
+    gradient_norm: int,
+    group=None,
+):
     """
     Gradient synch tolerant to missing grads.
 
-    Missing grads occur when some parameters are not trained between two
-    gradient synchs, e.g. the embeddings of a low-resource language with low
-    sampling weight.
+    Missing grads occur when some parameters are trained on some devices in
+    a communication group but not on others, between two gradient synchs.
 
-    The algorithm first uses the 'has_grad' attribute set by the forward hook
-    'has_grad_hook'. This hook ensures that all parameters of the modules
-    selected for use during the current training computation have 'has_grad'
-    set to True. This gives the list of parameters that have been trained on
-    this device ("ready").
+    The "managed" implementation relies on a deterministic sampling of tasks
+    known to all devices. This allows a device to figure out that even though
+    it didn't train some parameters itself, some other device did. In this case
+    the device must send a dummy gradient of all zeros.
 
-    A bit mask covering the parameters that are ready on this device is
-    communicated to the other devices in the group. The bit masks are reduced
-    using summation. The sum gives the number of real gradients for that
-    parameter, and can be used for normalization.
+    Only if no device trains some parameters (or if the parameters exist on
+    exactly one device) is it possible to skip communication entirely.
 
-    If a parameter is ready on any device, all devices communicate a value.
-    Devices on which the parameter is ready communicate the actual gradient,
-    while devices on which it is not ready communicate a dummy zero tensor
-        instead. The sum computed previously is used for normalization.
+    The "managed" implementation does not require the forward hook 'has_grad_hook'.
```

Comment:
good to know the hook can go, but bad for docstring. Move to a comment block outside the function / document this in the PR